General parameter scaling

jhod0 opened the pull request:
# General parameter scaling

This is an implementation of a feature I have discussed with Simon and Anowar. I added a few tests which pass but have not played with it too much yet.

It also includes a small refactor of how parameters are handled, which made it easier for me to implement this. The interface remains the same.

## Explanation

I am working on modeling lensing galaxy clusters and would like to fit the member mass profiles together in a uniform way, scaling their overall mass and radius by their magnitude. Something similar to these equations from [Bergamini 2019](https://ui.adsabs.harvard.edu/abs/2019A%26A...631A.130B/abstract):

![image](https://user-images.githubusercontent.com/8944001/186019141-2f8db310-7e84-4f39-8491-07cdb27b210a.png)

The existing `mass_scaling` parameter is similar but not exactly what I need; it only scales the overall mass linearly.

This PR implements generalized scaling, where any lens parameter can be scaled like this::

```
kwargs_lens[i]['param'] = param_scale * kwargs_lens[i]['param'] ** param_power
```

Where both `param_scale` and `param_power` can be fixed or sampled.

## Use

The interface is similar to `mass_scaling`, a mask is provided for each scaled parameter listing which lenses will be scaled.

This example will scale the `sigma0` and `Rs` parameters of `PJAFFE` profiles, like in the Bergamini example above. The member lens masses are labeled with `1`.

```
kwargs_constraints = {
    'general_scaling': {
        'sigma0': [False]*num_halo + [1]*nmembers,
        'Rs': [False]*num_halo + [1]*nmembers,
    }
}
```

And the initial values and bounds can be provided like this. This fixes the power-law exponent but varies the overall scaling for both `sigma0` and `Rs`.

```
fixed_special = {'sigma0_scale_pow': [alpha*2], 'Rs_scale_pow': [beta]}
kwargs_special_init = {'sigma0_scale_factor': [17.0], 'Rs_scale_factor': [8]}
kwargs_special_sigma = {'sigma0_scale_factor': [10.0], 'Rs_scale_factor': [3]}
kwargs_lower_special = {'sigma0_scale_factor': [0.5], 'Rs_scale_factor': [1]}
kwargs_upper_special = {'sigma0_scale_factor': [40], 'Rs_scale_factor': [20]}

special_params = [kwargs_special_init, kwargs_special_sigma, fixed_special, kwargs_lower_special, kwargs_upper_special]
```

sibirrer commented:
Hi @jhod0 , this looks great and an overall improvement of the style of parameter handing with a neat feature added.
The tests are failing because relative imports are apparently not supported to that level with '..', see error message.

- Can you add a test function for each class and give a description of the meaning of each class in words as doc strings?
- Is it possible for you to add about the same description for the use case as you nicely laid out in this issue in the doc strings of the Param() class? (or where you find it best suited)

Once the documentation of your current design is there, I have it easier to go through the code line-by-line for more minor things.

Thank you very much!

jhod0 commented:
Great, thanks Simon!

I have fixed the failing test and added docstrings for each of the new classes, and a blurb in the `Param` class similar to what I wrote in this post.

I have added a new test file for the abstract classes implemented in `Sampling/param_group.py`.

Let me know if you think this needs more documentation or other tweaks!

sibirrer commented on lenstronomy/PointSource/point_source_param.py:
```diff
@@ -2,10 +2,52 @@
+    _kwargs_lower = {'ra_image': -100, 'dec_image': -100, }
+    _kwargs_upper = {'ra_image': 100, 'dec_image': 100, }
+    def __init__(self, num_images):
+        self.on = int(num_images) > 0
```
can you document the variable num_images in doc strings?

sibirrer commented on lenstronomy/PointSource/point_source_param.py:
```diff
@@ -2,10 +2,52 @@
+    _kwargs_upper = {'source_amp': 100}
+
+
+class PointAmp(ArrayParam):
```
I would re-name this class to ImageAmp as it means the amplitude of the point source as seen in the image. I want to re-name this variable and there is actually a PR for this. But perhaps re-naming the class first is a step in the direction

sibirrer commented on lenstronomy/PointSource/point_source_param.py:
```diff
@@ -2,10 +2,52 @@
+
+class PointAmp(ArrayParam):
+    '''
+    Point amplification, possibly many
```
here the description should be 'amplitude of the point source as seen in the image' or something along these lines

sibirrer commented on lenstronomy/PointSource/point_source_param.py:
```diff
@@ -2,10 +2,52 @@
+    '''
+    _kwargs_lower = {'point_amp': 0}
+    _kwargs_upper = {'point_amp': 100}
+    def __init__(self, fixed_magnification):
```
please describe the variable fixed_magnification in the docstring

sibirrer commented on lenstronomy/PointSource/point_source_param.py:
```diff
@@ -32,36 +74,40 @@ def __init__(self, model_list, kwargs_fixed, num_point_source_list=None, linear_
-                kwargs_lower.append(fixed_low)
+            for model_params in self.param_groups:
+                fixed_lower = {}
+                for grp in model_params:
```
'grp' as a variable does not have much meaning for me. Can you re-name it to something a bit more obviously meaningful?

sibirrer commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,196 @@
+        '''
+        Tells the number of parameters that this group samples and theri names.
+
+        returns: 2-tuple of (num param, list of names)
```
the sphinx convention is :returns:

sibirrer commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,196 @@
+__all__ = ['ModelParamGroup', 'SingleParam', 'ArrayParam']
+
+
+class ModelParamGroup:
```
in this class, many input arguments are not documented. Use 
>>> :param <variable_name>: blabla
>>> :type <variable_name>: python type

sibirrer commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,196 @@
+        return output_kwargs, i
+
+
+class SingleParam(ModelParamGroup):
```
please make docstrings for input and output variables

sibirrer commented on lenstronomy/Sampling/parameters.py:
````diff
@@ -47,6 +47,42 @@ class that handles the parameter constraints. In particular when different model
+
+    ```
+    'general_scaling': {
+        'sigma0': [False]*num_halo + [1]*nmembers,
````
I would simplify the example. 'num_halo' and 'nmembers' are not really defined nor strictly necessary.
I guess currently there is no raise statement when the parameter in question (i.e. 'sigma0' in this example) is not part of the parameters for the model that is set to [True]. Perhaps make a TODO note that this is perhaps desired in the future

sibirrer commented on lenstronomy/Sampling/parameters.py:
````diff
@@ -47,6 +47,42 @@ class that handles the parameter constraints. In particular when different model
+    For example, if we want to jointly constrain the `sigma0` and `Rs` parameters
+    of some lens models, we can add:
+
+    ```
````
in the PR, you had a screen shot of the scaling law. Would you mind typing it down in latex here.

For docstrings, use

.. math::

  \\sigma_{0} = a^b

sibirrer replied:
make sure you use '\ \\' as one '\\' gets cut short

sibirrer commented on lenstronomy/Sampling/parameters.py:
```diff
@@ -77,6 +113,8 @@ def __init__(self, kwargs_model,
                  joint_source_with_source=[], joint_lens_with_light=[], joint_source_with_point_source=[],
                  joint_lens_light_with_point_source=[], joint_extinction_with_lens_light=[],
                  joint_lens_with_source_light=[], mass_scaling_list=None, point_source_offset=False,
+                 # General scaling: need names of params and number of individual params
```
can you add the description of the parameter further down with :param general_scaling:?

sibirrer commented on lenstronomy/Sampling/parameters.py:
```diff
@@ -544,24 +591,41 @@ def update_lens_scaling(self, kwargs_special, kwargs_lens, inverse=False):
-                elif 'k_eff' in kwargs:
-                    kwargs['k_eff'] *= scale_factor
+
+        if self._mass_scaling:
```
we could think of even fully removing mass_scaling and replace it with something within the power_law scaling that automatically sets the power=1. You don't need to do it, but perhaps mark a TODO item in the code, or even a DeprecationWarning that this will come.

jhod0 replied:
that makes a lot of sense, I thought about having the general_scaling default the power law index to 1 as well. I decided not to because it could cause confusion, if for instance someone provides the power law index with a typo and lenstronomy silently defaults to 1. I think reimplementing mass_scaling to use the same machinery as general_scaling makes most sense. I am adding a TODO along these lines

sibirrer replied:
ok, we can leave it like the and thanks for marking it as TODO

sibirrer commented on lenstronomy/Sampling/parameters.py:
```diff
@@ -642,6 +706,7 @@ def print_setting(self):
         num, param_list = self.num_param()
         num_linear = self.num_param_linear()
 
+        # TODO print settings of specailParams?
```
Yes, if you want to, that would be great. I have neglected this print statements a bit...

jhod0 replied:
I am leaving this as a TODO because I am not sure how best to display the information. the special params grouping is a bit eclectic and I am not sure what all of the options mean

sibirrer replied:
I was just thinking you add the print statement of the option you are implementing. I don't want to blame you for code that is already there, just want to make sure that new features being implemented get added and generally improve the usability

sibirrer reviewed (changes requested):
All in all looks very good. I am a bit stingy with documentation (more and more with every added feature) also such that the documentation looks good. You can test out the sphinx auto-generated documentation with 
>>> $make docs
(when you are in the base lenstronomy repo where the make file is and you have installed the sphinx extensions)

I would like another review by @ajshajib  to follow best practices with multiply eyes on the code (you are the ginny pig here...)

sibirrer commented:
oh, and add your name and link to your GitHub page in the AUTHORS.rst file :)

sibirrer commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,250 @@
+
+        return npar, names
+
+    def set_params(self, kwargs, kwargs_fixed):
```
there are still a few routines without doc strings...sorry for being pedantic :)

sibirrer reviewed (commented):
Looks very good and thank you very much for the added documentation!
I have some very minor comments left and that's it from my side. I leave it then for Anowar to glance over it in case

jhod0 commented:
No problem, I appreciate the extra care to make lenstronomy great!

For some reason I can't reply above, but I think `fixed_magnification` is an integer. This is a snippet of how it was implemented before, where the `point_amp` parameter is set to an array.

Maybe I am misunderstanding what the intention was. In any case, the array was the reason I implemented `ImageAmp` as an `ArrayParam` subclass

https://github.com/lenstronomy/lenstronomy/blob/1e6d9311bed4e4af19f62df965e8d9ad4fc54f0e/lenstronomy/PointSource/point_source_param.py#L45-L48

sibirrer commented:
> No problem, I appreciate the extra care to make lenstronomy great!
> 
> For some reason I can't reply above, but I think `fixed_magnification` is an integer. This is a snippet of how it was implemented before, where the `point_amp` parameter is set to an array.
> 
> Maybe I am misunderstanding what the intention was. In any case, the array was the reason I implemented `ImageAmp` as an `ArrayParam` subclass
> 
> https://github.com/lenstronomy/lenstronomy/blob/1e6d9311bed4e4af19f62df965e8d9ad4fc54f0e/lenstronomy/PointSource/point_source_param.py#L45-L48

'fixed_magnification' is a list of booleans for each point source model to state whether the image amplitudes are fixed in respect to the source amplitude through the lensing magnification, or instead treated as stand-alone independent amplitude parameters. If the argument is true, it uses one 'source_amp' parameter, and if the argument is False, it uses the length of image positions as the argument.

jhod0 commented:
OK, I have confused myself a little, and I was using the fixed_magnification correctly, but here is my understanding:

- if the magnification is fixed, a single parameter `source_amp` is sampled per point source model
  - this corresponds to the SourceAmp class I have defined
- if the magnification is not fixed, `num_point_sources` are sampled, one for each point source
  - this corresponds to the ImageAmp class I have defined

I think my implementation was correct but the naming was wrong. I have changed the name and documentation accordingly

https://github.com/lenstronomy/lenstronomy/blob/1e6d9311bed4e4af19f62df965e8d9ad4fc54f0e/lenstronomy/PointSource/point_source_param.py#L102-L113

https://github.com/lenstronomy/lenstronomy/blob/1e6d9311bed4e4af19f62df965e8d9ad4fc54f0e/lenstronomy/PointSource/point_source_param.py#L180-L188

sibirrer reviewed (approved):
Great, I approve this PR. @ajshajib  should also have a look at it.

ajshajib commented on AUTHORS.rst:
```diff
@@ -34,7 +35,7 @@ Contributors (alphabetic)
 * Brian Nord `bnord <https://github.com/bnord/>`_
 * Giulia Pagano
 * Ji Won Park `jiwoncpark <https://github.com/jiwoncpark/>`_
-* Andreas Filipp `andreasfilipp <https://github.com/andreasfilipp/>`_
```
Why was Andreas Filipp removed?

sibirrer replied:
I think he was moved as not alphabetically ordered

ajshajib replied:
O, alright.

jhod0 replied:
I just moved them up in the list to keep alphabetic order! Certainly not trying to remove their credit haha. Andreas now appears after Sydney Erickson

ajshajib commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,321 @@
+    :param _kwargs_lower: Dictionary. Lower bounds of each parameter
+    :param _kwargs_upper: Dictionary. Upper bounds of each parameter
+    '''
+    def __init__(self, on):
```
I suggest to add a docstring for `on`. Also, if `self.on` is only intended to be used within this class instance itself, I recommend to rename the variable to `self._on`, `_` flags that it's a "private" variable.

jhod0 replied:
I have refactored this as a read-only variable like you suggested. All the subclass initializers changed accordingly

ajshajib commented on lenstronomy/Sampling/param_group.py:
```diff
@@ -0,0 +1,321 @@
+    '''
+    def num_params(self, kwargs_fixed):
+        '''
+        Tells the number of parameters that this group samples and theri names.
```
theri → there

jhod0 replied:
done, thanks!

ajshajib reviewed (approved):
This looks good to me. Thanks @jhod0 for making the requested changes, and also for making this very helpful PR. I am looking forward to using this for some of my group-scale lenses. 😊
